python/perspective/bench/bench.py

```diff
@@ -0,0 +1,200 @@
+        """
+        raise NotImplementedError(
+                "Must implement `register_benchmarks` to run benchmark suite.")
+
+    def before_all(self):
+        pass
+
+    def after_all(self):
+        pass
+
+    def before_each(self):
+        pass
+
+    def after_each(self):
+        pass
+
+
+class Runner(object):
+
+    ITERATIONS = 10
+
+    def __init__(self, suite):
+        """Initializes a benchmark runner for the `Suite`.
+
+        Args:
+            suite (Suite) : A class that inherits from `Suite`, with any number
+                of instance methods decorated with `@benchmark`.
+        """
+        self._suite = suite
+        self._benchmarks = []
+        self._table = None
+        self._WROTE_RESULTS = False
+        self._HOSTING = False
+
+        self._suite.register_benchmarks()
+
+        class_attrs = self._suite.__class__.__dict__.items()
+        instance_attrs = self._suite.__dict__.items()
+
+        for (k, v) in class_attrs:
+            if hasattr(v, "benchmark") and getattr(v, "benchmark") is True:
+                logging.info("Registering {0}".format(k))
+                self._benchmarks.append(v)
+
+        for (k, v) in instance_attrs:
+            if hasattr(v, "benchmark") and getattr(v, "benchmark") is True:
+                logging.info("Registering {0}".format(k))
+                self._benchmarks.append(v)
+
+        # Write results on SIGINT
+        signal.signal(signal.SIGINT, self.sigint_handler)
+
+    def sigint_handler(self, signum, frame):
+        """On SIGINT, host the results over a websocket."""
+        if not self._WROTE_RESULTS:
+            self.write_results()
+        if not self._HOSTING:
+            self.host_results()
+        else:
+            sys.exit(0)
+
+    def host_results(self):
+        """Create a tornado application that hosts the results table over a
+        websocket."""
+        if self._table is None:
+            return
+        MANAGER = PerspectiveManager()
+        MANAGER.host_table("benchmark_results", self._table)
+        application = tornado.web.Application([
+            (r"/", BenchmarkTornadoHandler),
+            # create a websocket endpoint that the client Javascript can access
+            (r"/websocket", PerspectiveTornadoHandler, {
+                "manager": MANAGER,
+                "check_origin": True
+            })
+        ])
+        self._HOSTING = True
+        application.listen(8888)
+        logging.critical("Displaying results at http://localhost:8888")
+        loop = tornado.ioloop.IOLoop.current()
+        loop.start()
+
+    def write_results(self):
+        if self._table is None:
+            return
+        logging.info("Writing results to `benchmark.arrow`")
+        arrow_path = os.path.join(os.path.dirname(__file__), "benchmark.arrow")
+        with open(arrow_path, "wb") as file:
+            arrow = self._table.view().to_arrow()
+            file.write(arrow)
+        self._WROTE_RESULTS = True
```

Comment:
It would be nice if this method could incrementally write to an existing arrow so we can locally archive multiple versions, by checking to see if an `benchmark.arrow` file exists on load.  It would also enable cross-language benchmarks in the same archive ..